src/x86_64/unix_machine.c

```diff
@@ -136,15 +138,19 @@ boolean setup_sigframe(thread t, int signum, struct siginfo *si)
         return false;
     void *fpstate;
     rsp = setup_ucontext_fpstate(&fpstate, t, rsp);
+    if (rsp == INVALID_PHYSICAL)
+        return false;
 
     /* Create space for rt_sigframe. Note that we are actually aligning to 8
        but not 16 bytes; the ABI requires that stacks are aligned to 16 before
        a call, but the sigframe return into the function takes the place of a
        call, which would have pushed a return address. The function prologue
        typically pushes the frame pointer on the stack, thus re-aligning to 16
        before executing the function body. */
-    rsp -= pad(sizeof(struct rt_sigframe), 16) + 8; /* ra offset */
-
+    u64 sz = pad(sizeof(struct rt_sigframe), 16) + 8; /* ra offset */
+    rsp -= sz;
+    if (!fault_in_user_memory_for_process(t->p, (void *)rsp, sz, true))
```

Comment:
It seems to me there is no need to to fault in explicitly the user memory, because we already have the `if (context_set_err(ctx)) return false;` instructions which ensure setup_sigframe() returns false if a page cannot be faulted in due to OOM conditions. So a call to fault_in_memory() would effectively do the same thing that's already being done by writing to the user stack page, e.g. in setup_ucontext(). The same goes for the aarch64 and riscv code.
I tried running the e2e tests after reverting this commit, and I couldn't reproduce the problem you describe where the VM gets stuck forever in a loop: e.g. the ruby_alloc test ends when either setup_sigframe() returns false (which as far as I can tell works correctly with the existing code), or a sequence of out-of-memory page faults occurs (in which case the new OOM kiiller takes care of shutting down the program).

Reply:
When I first started looking at this, before I had the OOM killer written, I was seeing two different symptoms related to OOM (after adding additional debugging in), and this fix eliminated one of the symptoms. I also can't seem to reproduce what I was seeing before. I see the context_set_err should take care of it in exactly the same way with or without this fix, so now I'm questioning what I saw originally. In any case, you are right and this fix isn't needed, so I dropped the commit. Thanks!